torchdata/nodes/adapters.py

```diff
@@ -34,33 +34,37 @@ class IterableWrapper(BaseNode[T]):
     ITERABLE_KEY = "iterable"
 
     def __init__(self, iterable: Iterable[T]):
+        super().__init__()
         self.iterable = iterable
-        self._num_yielded = 0
+        self._it: Optional[Iterator[T]] = None
 
-    def iterator(self, initial_state: Optional[Dict[str, Any]]) -> Iterator[T]:
+    def reset(self, initial_state: Optional[Dict[str, Any]] = None):
         self._num_yielded = 0
+        self._it = None
+        super().reset(initial_state)
         if initial_state is not None:
             self._num_yielded = initial_state[self.NUM_YIELDED_KEY]
             if isinstance(self.iterable, Stateful):
                 self.iterable.load_state_dict(initial_state[self.ITERABLE_KEY])
-                it = iter(self.iterable)
+                self._it = iter(self.iterable)
             else:
-                it = iter(self.iterable)
+                self._it = iter(self.iterable)
                 # Naively fast-forwarding
                 for i in range(self._num_yielded):
                     try:
-                        next(it)
+                        next(self._it)
                     except StopIteration:
                         raise ValueError(
                             f"Tried to fast-forward {self._num_yielded} items during init but "
                             f"hit StopIteration after {i} items, this is likely a bug or malformed state_dict"
                         )
         else:
-            it = iter(self.iterable)
+            self._it = iter(self.iterable)
 
-        for item in it:
-            self._num_yielded += 1
-            yield item
+    def next(self) -> T:
+        item = next(self._it)  # type: ignore [arg-type, union-attr]
```

Comment:
There might be a more-clever way to handle this, eg through a property or something, that won't make the type-hinters complain. Would prefer to do it in a separate diff as it's currently mostly sugar. Alternatively we can also `assert self._it is not None`